server(actions): add support for queries (close #4032) (#4309)
hasura/graphql-engine: docs/graphql/manual/actions/index.rst

What are actions?
-----------------

> Actions are a way to extend Hasura's schema with entirely

Reviewer: ```suggestion
Actions are a way to extend Hasura's schema with custom business logic using custom queries and mutations.
```

> custom queries or mutations with custom business logic. Actions can be

Reviewer: ```suggestion
Actions can be
```

An action consists of the following parts:

> 1. ``Type``: The type of the action (``query`` or ``mutation``).

Reviewer: ```suggestion
1. ``Type``: The type of the action (``query`` or ``mutation``)
```

> 2. ``Definition``: The definition of the query/mutation

Reviewer: ```suggestion
2. ``Definition``: The definition of the query or mutation
```

> 2. ``Definition``: The definition of the query/mutation

Reviewer: ```suggestion
2. ``Definition``: The definition of the query mutation
```

3. ``Handler``: The logic to be run when the query/mutation is executed

> 4. ``Kind``: Sync or async.

Reviewer: ```suggestion
4. ``Kind``: Sync or async
```

> NOTE: In case of a Query Action, there is no ``Kind`` associated with the action.

Reviewer: ```suggestion
.. note::
	In case of a query action, there is no ``Kind`` associated with the action.
```

> A query action works like a ``Synchronous`` Mutation Action.

Reviewer: ```suggestion
	A query action works like a ``Synchronous`` mutation action.
```

Type
****

> Actions are of two types:

Reviewer: ```suggestion
Actions can be of two types:
```

> - **Query Action**: An action of type ``query`` extends the query root of the Hasura schema. This means that you can execute this action through a GraphQL query. Query actions must be used where you want to fetch data from a data source without changing anything on the data source.

Reviewer: ```suggestion
- **Query action**: An action of type ``query`` extends the query root of the Hasura schema. This means that you can execute this action through a GraphQL query. Query actions must be used where you want to fetch data from a data source without changing anything on the data source.
```

> - **Mutation Action**: An action of type ``mutation`` extends the mutation root of the Hasura schema. This means that you can execute this action through a GraphQL mutation. Mutation actions must be used when you want to mutate the state of a data source and fetch some data.

Reviewer: ```suggestion
- **Mutation action**: An action of type ``mutation`` extends the mutation root of the Hasura schema. This means that you can execute this action through a GraphQL mutation. Mutation actions must be used when you want to mutate the state of a data source and fetch some data.
```

Definition
**********

The action definition consists of the following:

> - ``Action Name``: The action will be available as a query/mutation in the GraphQL

Reviewer: ```suggestion
- ``Action name``: The action will be available as a query or mutation in the GraphQL
```

  schema named as the action name
- ``Arguments``: Arguments are used to pass dynamic values along with the

>   query/mutation.

Reviewer: ```suggestion
  query or mutation.
```

> - ``Response type``: The GraphQL type of the response that the query/mutation will

Reviewer: ```suggestion
- ``Response type``: The GraphQL type of the response that the query or mutation will
```

Kind
****

> Mutation Actions are of two kinds:

Reviewer: ```suggestion
Mutation actions are of two kinds:
```

  handler and allow the client to subscribe to the actual response using the
  ``action id``.

> Query Actions don't have a kind, they always behave like ``Synchronous Mutation Actions``.

Reviewer: ```suggestion
Query actions don't have a kind, they always behave like ``Synchronous mutation actions``.
```

How it works?
-------------

> * Hasura receives the action GraphQL query/mutation and converts this request into an

Reviewer: ```suggestion
* Hasura receives the action GraphQL query or mutation and converts this request into an
```
